File: apps/ios/Loonext/Features/Thread/ContactPanelSheet.swift
import SwiftUI

/// The thread's contact panel — the web sidebar as a sheet, opened by tapping
/// the header identity (the Android ContactPanelSheet.kt twin): inline
/// name/address/notes with the G6 800ms auto-save, the consent line, the
/// conversation's open-tasks checklist (T5.2 — done toggles through the
/// source message, never a task route), and prior conversations with this
/// contact (rows route through the caller into the inbox's open command).
@MainActor
struct ContactPanelSheet: View {
    let controller: ThreadController
    let members: [Member]
    /// Navigation into ANOTHER thread; rows stay un-tappable until it's wired
    /// (a row that goes nowhere would be a lie).
    let onOpenConversation: (@MainActor (String) -> Void)?
    /// Open a checklist task's detail (#217). The caller dismisses this sheet
    /// and routes the task up to the shell.
    let onOpenTask: @MainActor (String) -> Void
    /// #465: open the FULL contact screen — its history, call log and every
    /// conversation. This panel holds a copy of the fields and nothing else, so
    /// without it the panel is a dead end, which is exactly what was reported.
    /// Web has had the jump since #82. The caller dismisses this sheet first,
    /// the same way onOpenTask and onOpenConversation do.
    let onOpenContact: @MainActor (String) -> Void

    @Environment(\.appLocale) private var appLocale

    var body: some View {
        Group {
            if let detail = controller.conversation {
                panelBody(detail)
            }
        }
        .presentationDetents([.medium, .large])
        .task { controller.loadContactPanel() }
    }

    private func memberName(_ userId: String?) -> String? {
        guard let userId else { return nil }
        let name = members.first { $0.user_id == userId }?.display_name
        return (name?.isBlank ?? true) ? nil : name
    }

    private func panelBody(_ detail: ConversationDetail) -> some View {
        let contact = controller.contact
        let displayName = detail.contact.name ?? formatPhone(detail.contact.phone_e164)
        return ScrollView {
            VStack(alignment: .leading, spacing: 14) {
                HStack(spacing: 12) {
                    InitialsAvatar(name: displayName, size: 44)
                    VStack(alignment: .leading, spacing: 2) {
                        Text(displayName)
                            .font(.golos(16, weight: .semibold))
                            .foregroundStyle(BrandColor.ink)
                            .lineLimit(1)
                        let phone = formatPhone(detail.contact.phone_e164)
                        let optedOut = AppStrings.translate(
                            appLocale, "thread.optedOut"
                        )
                        Text(
                            contact?.opted_out == true
                                ? "\(phone) · \(optedOut)"
                                : phone
                        )
                        .font(.golos(11))
                        .foregroundStyle(BrandColor.muted500)
                    }
                    Spacer(minLength: 8)
                    Button {
                        onOpenContact(detail.contact_id)
                    } label: {
                        Image(systemName: "arrow.up.forward.square")
                            .font(.scaled(17, weight: .regular))
                            .foregroundStyle(BrandColor.muted500)
                    }
                    .buttonStyle(.plain)
                    .accessibilityLabel(
                        AppStrings.translate(appLocale, "thread.openFullContact")
                    )
                }
                .padding(.top, 18)

                // Details — the same auto-saving fields as the contact screen.
                sheetSection(
                    AppStrings.translate(appLocale, "thread.sectionDetails")
                ) {
                    PanelAutosaveField(
                        label: AppStrings.translate(appLocale, "thread.fieldName"),
                        initial: contact?.name ?? detail.contact.name ?? "",
                        maxLength: contactNameMax,
                        placeholder: AppStrings.translate(
                            appLocale, "thread.addName"
                        ),
                        multiline: false
                    ) { value in
                        try await controller.saveContactField("name", value)
                    }
                    .id("\(detail.contact_id)|name")
                    PanelAutosaveField(
                        label: AppStrings.translate(
                            appLocale, "thread.fieldAddress"
                        ),
                        initial: contact?.address ?? detail.contact.address ?? "",
                        maxLength: contactAddressMax,
                        placeholder: AppStrings.translate(
                            appLocale, "thread.addAddress"
                        ),
                        multiline: false
                    ) { value in
                        try await controller.saveContactField("address", value)
                    }
                    .id("\(detail.contact_id)|address")
                    PanelAutosaveField(
                        label: AppStrings.translate(appLocale, "thread.fieldNotes"),
                        initial: contact?.notes ?? detail.contact.notes ?? "",
                        maxLength: contactNotesMax,
                        placeholder: AppStrings.translate(
                            appLocale, "thread.notesPlaceholder"
                        ),
                        multiline: true
                    ) { value in
                        try await controller.saveContactField("notes", value)
                    }
                    .id("\(detail.contact_id)|notes")
                }

                sheetSection(
                    AppStrings.translate(appLocale, "thread.sectionConsent")
                ) {
                    Text(
                        consentLine(
                            consentSource: contact?.consent_source
                                ?? detail.contact.consent_source,
                            consentAt: contact?.consent_at ?? detail.contact.consent_at,
                            consentAttestedBy: contact?.consent_attested_by,
                            memberName: memberName,
                            locale: appLocale
                        )
                    )
                    .font(.subheadline)
                    .foregroundStyle(.secondary)
                }

                // #301: where this customer came from. Above the tasks
                // because it is a question somebody ASKS in the first minute
                // of a call, while a task is what they write down afterwards —
                // and because the ask disappears the moment it is answered.
                sheetSection(
                    AppStrings.translate(appLocale, "thread.sectionLeadSource")
                ) {
                    LeadSourcePicker(controller: controller, detail: detail)
                }

                sheetSection(
                    AppStrings.translate(appLocale, "thread.sectionTasks")
                ) {
                    TasksChecklistSection(
                        state: controller.conversationTasks,
                        onToggle: { controller.toggleTaskDone($0) },
                        onOpenTask: onOpenTask
                    )
                }

                sheetSection(
                    AppStrings.translate(
                        appLocale, "thread.sectionOtherConversations"
                    )
                ) {
                    OtherConversationsSection(
                        state: controller.otherConversations,
                        onOpen: onOpenConversation
                    )
                }
            }
            .padding(.horizontal, 20)
            .padding(.bottom, 24)
            .frame(maxWidth: .infinity, alignment: .leading)
        }
        .background(BrandColor.canvas.ignoresSafeArea())
    }

    private func sheetSection(
        _ title: String,
        @ViewBuilder content: () -> some View
    ) -> some View {
        VStack(alignment: .leading, spacing: 6) {
            SectionHeader(label: title)
            content()
        }
        .frame(maxWidth: .infinity, alignment: .leading)
    }
}

// MARK: - Sections

/// The conversation checklist (T5.2). Done toggles ride the SOURCE MESSAGE's
/// done bit — the controller enforces the invariant.
@MainActor
private struct TasksChecklistSection: View {
    let state: LoadState<[TaskItem]>?
    let onToggle: @MainActor (TaskItem) -> Void
    /// #217: tapping the ROW (not the checkbox) opens the task's detail.
    let onOpenTask: @MainActor (String) -> Void

    @Environment(\.appLocale) private var appLocale

    var body: some View {
        switch state {
        case nil, .loading?:
            ProgressView()
        case .failed?:
            Text(AppStrings.translate(appLocale, "thread.tasksLoadFailed"))
                .font(.footnote)
                .foregroundStyle(.secondary)
        case .ready(let tasks)?:
            if tasks.isEmpty {
                Text(AppStrings.translate(appLocale, "thread.noTasks"))
                    .font(.footnote)
                    .foregroundStyle(.secondary)
            } else {
                VStack(alignment: .leading, spacing: 0) {
                    ForEach(tasks, id: \.id) { task in
                        taskRow(task)
                    }
                }
            }
        }
    }

    /// Two separate hit targets (#217): the checkbox toggles done through the
    /// source message (never navigates), the title area opens the task detail.
    private func taskRow(_ task: TaskItem) -> some View {
        HStack(spacing: 10) {
            Button {
                onToggle(task)
            } label: {
                Image(systemName: task.done ? "checkmark.square.fill" : "square")
                    .foregroundStyle(
                        task.done
                            ? AnyShapeStyle(BrandColor.olive)
                            : AnyShapeStyle(BrandColor.muted250)
                    )
                    .contentShape(Rectangle())
            }
            .buttonStyle(.plain)
            .accessibilityLabel(
                AppStrings.translate(
                    appLocale,
                    task.done
                        ? "thread.reopenTaskNamed"
                        : "thread.markTaskDoneNamed",
                    ["title": task.title]
                )
            )

            Button {
                onOpenTask(task.id)
            } label: {
                HStack(spacing: 0) {
                    Text(task.title)
                        .font(.subheadline)
                        .strikethrough(task.done)
                        .foregroundStyle(task.done ? .secondary : .primary)
                        .lineLimit(2)
                        .multilineTextAlignment(.leading)
                    Spacer(minLength: 0)
                }
                .contentShape(Rectangle())
            }
            .buttonStyle(.plain)
            .accessibilityLabel(
                AppStrings.translate(
                    appLocale, "thread.openTaskNamed", ["title": task.title]
                )
            )
        }
        .padding(.vertical, 6)
    }
}

/// Prior conversations with this contact (the current thread excluded).
@MainActor
private struct OtherConversationsSection: View {
    let state: LoadState<[ConversationListItem]>?
    let onOpen: (@MainActor (String) -> Void)?

    @Environment(\.appLocale) private var appLocale

    var body: some View {
        switch state {
        case nil, .loading?:
            ProgressView()
        case .failed?:
            Text(AppStrings.translate(appLocale, "thread.priorLoadFailed"))
                .font(.footnote)
                .foregroundStyle(.secondary)
        case .ready(let rows)?:
            if rows.isEmpty {
                Text(
                    AppStrings.translate(
                        appLocale, "thread.noOtherConversations"
                    )
                )
                .font(.footnote)
                .foregroundStyle(.secondary)
            } else {
                VStack(alignment: .leading, spacing: 0) {
                    ForEach(rows, id: \.id) { row in
                        let content = HStack(alignment: .center, spacing: 8) {
                            VStack(alignment: .leading, spacing: 1) {
                                Text(snippetLabel(row))
                                    .font(.subheadline)
                                    .foregroundStyle(.primary)
                                    .lineLimit(1)
                                Text(statusLabel(row.status, locale: appLocale))
                                    .font(.caption)
                                    .foregroundStyle(.secondary)
                            }
                            Spacer()
                            Text(relativeTime(row.last_message_at))
                                .font(.caption)
                                .foregroundStyle(.secondary)
                        }
                        .padding(.vertical, 8)
                        if let onOpen {
                            Button {
                                onOpen(row.id)
                            } label: {
                                content
                            }
                            .buttonStyle(.plain)
                        } else {
                            content
                        }
                        Divider()
                    }
                }
            }
        }
    }

    private func snippetLabel(_ row: ConversationListItem) -> String {
        guard let body = row.last_message?.body, !body.isBlank else {
            return AppStrings.translate(appLocale, "thread.conversationFallback")
        }
        return body
    }
}

// MARK: - Auto-save field

/// G6 auto-save, the contact detail screen's field verbatim (that one is file-
/// private to ContactDetailView.swift): writes the field 800ms after the last
/// keystroke (blank clears — an explicit null on the wire) with a quiet
/// status line. A new keystroke during a pending save restarts the clock; the
/// newest value wins.
@MainActor
private struct PanelAutosaveField: View {
    private enum SaveState {
        case idle, saving, saved, failed
    }

    let label: String
    let initial: String
    let maxLength: Int
    let placeholder: String
    let multiline: Bool
    let save: (String?) async throws -> Void

    @State private var value: String
    @State private var lastSaved: String
    @State private var saveState: SaveState = .idle

    @Environment(\.appLocale) private var appLocale

    init(
        label: String,
        initial: String,
        maxLength: Int,
        placeholder: String,
        multiline: Bool,
        save: @escaping (String?) async throws -> Void
    ) {
        self.label = label
        self.initial = initial
        self.maxLength = maxLength
        self.placeholder = placeholder
        self.multiline = multiline
        self.save = save
        _value = State(initialValue: initial)
        _lastSaved = State(initialValue: initial)
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 2) {
            Text(label)
                .font(.caption)
                .foregroundStyle(.secondary)
            TextField(placeholder, text: $value, axis: multiline ? .vertical : .horizontal)
                .font(.subheadline)
                .lineLimit(multiline ? 3 ... 6 : 1 ... 1)
                .onChange(of: value) { _, next in
                    if next.count > maxLength {
                        value = String(next.prefix(maxLength))
                    }
                }
            Text(statusLine)
                .font(.caption)
                .foregroundStyle(
                    saveState == .failed
                        ? AnyShapeStyle(BrandColor.destructive)
                        : AnyShapeStyle(Color.secondary)
                )
                .frame(height: 14)
        }
        .task(id: value) {
            let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
            let savedTrimmed = lastSaved.trimmingCharacters(in: .whitespacesAndNewlines)
            guard trimmed != savedTrimmed else { return }
            // The debounce: a new keystroke cancels this task and starts over.
            try? await Task.sleep(for: .milliseconds(800))
            if Task.isCancelled { return }
            saveState = .saving
            do {
                try await save(trimmed.isEmpty ? nil : trimmed)
                lastSaved = value
                saveState = .saved
            } catch {
                saveState = .failed
            }
        }
    }

    private var statusLine: String {
        switch saveState {
        case .idle: ""
        case .saving: AppStrings.translate(appLocale, "common.saving")
        case .saved: AppStrings.translate(appLocale, "common.saved")
        case .failed: AppStrings.translate(appLocale, "thread.saveFailed")
        }
    }
}

// MARK: - Previews

#Preview("Panel sections") {
    ScrollView {
        VStack(alignment: .leading, spacing: 14) {
            TasksChecklistSection(
                state: .ready([
                    TaskItem(
                        id: "t1",
                        company_id: "co",
                        message_id: "m1",
                        conversation_id: "c1",
                        title: "Send the fence quote",
                        description: "",
                        assigned_user_id: nil,
                        due_at: nil,
                        created_by_user_id: "u1",
                        created_at: "2026-07-15T12:00:00Z",
                        updated_at: "2026-07-15T12:00:00Z",
                        done: false,
                        status: "open",
                        contact: nil,
                        attachment_count: 0
                    ),
                    TaskItem(
                        id: "t2",
                        company_id: "co",
                        message_id: "m2",
                        conversation_id: "c1",
                        title: "Confirm the gate code",
                        description: "",
                        assigned_user_id: nil,
                        due_at: nil,
                        created_by_user_id: "u1",
                        created_at: "2026-07-14T12:00:00Z",
                        updated_at: "2026-07-15T12:00:00Z",
                        done: true,
                        status: "done",
                        contact: nil,
                        attachment_count: nil
                    ),
                ]),
                onToggle: { _ in },
                onOpenTask: { _ in }
            )
            Divider()
            OtherConversationsSection(
                state: .ready([
                    ConversationListItem(
                        id: "c2",
                        company_id: "co",
                        contact_id: "p1",
                        phone_number_id: "n1",
                        status: "closed",
                        is_spam: false,
                        assigned_user_id: nil,
                        pinned_at: nil,
                        pinned_by_user_id: nil,
                        last_message_at: "2026-07-10T09:00:00Z",
                        closed_at: "2026-07-11T09:00:00Z",
                        created_at: "2026-07-09T09:00:00Z",
                        updated_at: "2026-07-11T09:00:00Z",
                        contact: ContactSummary(
                            id: "p1",
                            name: "Dana Whitcomb",
                            phone_e164: "+14155550134"
                        ),
                        tags: [],
                        unread: false,
                        last_message: ConversationSnippet(
                            id: "m9",
                            direction: "outbound",
                            body: "Thanks — see you Tuesday at 9.",
                            created_at: "2026-07-10T09:00:00Z",
                            has_attachments: false
                        )
                    ),
                ]),
                onOpen: { _ in }
            )
        }
        .padding(20)
    }
}

File: apps/android/app/src/main/kotlin/com/loonext/android/features/thread/ContactPanelSheet.kt
package com.loonext.android.features.thread

import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.material3.FilterChip
import androidx.compose.material3.TextButton
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import com.loonext.android.core.i18n.t
import com.loonext.android.core.model.ConversationDetail
import com.loonext.android.core.model.LeadSource
import kotlinx.coroutines.launch
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.outlined.OpenInNew
import androidx.compose.material3.Checkbox
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.LoadingIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.loonext.android.core.model.Member
import com.loonext.android.core.model.Task
import com.loonext.android.features.contacts.AutosaveField
import com.loonext.android.features.contacts.CONTACT_ADDRESS_MAX
import com.loonext.android.features.contacts.CONTACT_NAME_MAX
import com.loonext.android.features.contacts.CONTACT_NOTES_MAX
import com.loonext.android.features.contacts.consentLine
import com.loonext.android.ui.common.AppSheet
import com.loonext.android.ui.common.InitialsAvatar
import com.loonext.android.ui.common.LoadState
import com.loonext.android.ui.common.formatPhone
import com.loonext.android.ui.common.relativeTime

/**
 * The thread's contact panel (#165) — the web sidebar as a bottom sheet,
 * opened by tapping the header identity: inline name/address/notes with the
 * G6 800ms auto-save (the exact field the contact detail screen uses), the
 * consent attest line, prior conversations with this contact, and the
 * conversation's open-tasks checklist (T5.2 — done toggles through the
 * source message, never a task route).
 *
 * [onOpenConversation] is the caller's navigation into ANOTHER thread; rows
 * stay un-tappable until it's wired (a row that goes nowhere would be a lie).
 */
@Composable
internal fun ContactPanelSheet(
    controller: ThreadController,
    members: List<Member>,
    onOpenConversation: ((conversationId: String) -> Unit)?,
    onOpenTask: ((taskId: String) -> Unit)?,
    /**
     * #465: open the FULL contact screen — its history, call log and every
     * conversation. This panel holds a copy of the fields and nothing else, so
     * without this it is a dead end, which is exactly what was reported. Web
     * has had the jump since #82.
     */
    onOpenContact: ((contactId: String) -> Unit)?,
    onDismiss: () -> Unit,
) {
    val detail = controller.conversation ?: return
    val contact = controller.contact
    val displayName = detail.contact.name ?: formatPhone(detail.contact.phone_e164)

    LaunchedEffect(controller) { controller.loadContactPanel() }

    fun memberName(userId: String?): String? =
        members.firstOrNull { it.user_id == userId }?.display_name?.ifBlank { null }

    AppSheet(onDismissRequest = onDismiss) {
        Column(
            Modifier
                .fillMaxWidth()
                .verticalScroll(rememberScrollState())
                .padding(horizontal = 20.dp),
            verticalArrangement = Arrangement.spacedBy(14.dp),
        ) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                InitialsAvatar(displayName, size = 44.dp)
                Spacer(Modifier.width(12.dp))
                Column(Modifier.weight(1f)) {
                    Text(
                        displayName,
                        style = MaterialTheme.typography.titleMedium,
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis,
                    )
                    Text(
                        if (contact?.opted_out == true) {
                            "${formatPhone(detail.contact.phone_e164)} · " +
                                t("thread.optedOut")
                        } else {
                            formatPhone(detail.contact.phone_e164)
                        },
                        style = MaterialTheme.typography.labelSmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                }
                if (onOpenContact != null) {
                    IconButton(
                        onClick = {
                            onDismiss()
                            onOpenContact(detail.contact_id)
                        },
                    ) {
                        Icon(
                            Icons.AutoMirrored.Outlined.OpenInNew,
                            contentDescription = t("thread.openFullContact"),
                            tint = MaterialTheme.colorScheme.onSurfaceVariant,
                            modifier = Modifier.size(19.dp),
                        )
                    }
                }
            }

            // Details — the same auto-saving fields as the contact screen.
            SheetSection(t("thread.sectionDetails")) {
                AutosaveField(
                    fieldKey = "${detail.contact_id}:name",
                    label = t("thread.fieldName"),
                    initial = (contact?.name ?: detail.contact.name).orEmpty(),
                    maxLength = CONTACT_NAME_MAX,
                    placeholder = t("thread.addName"),
                    singleLine = true,
                    save = { value -> controller.saveContactField("name", value) },
                )
                AutosaveField(
                    fieldKey = "${detail.contact_id}:address",
                    label = t("thread.fieldAddress"),
                    initial = (contact?.address ?: detail.contact.address).orEmpty(),
                    maxLength = CONTACT_ADDRESS_MAX,
                    placeholder = t("thread.addAddress"),
                    singleLine = true,
                    save = { value -> controller.saveContactField("address", value) },
                )
                AutosaveField(
                    fieldKey = "${detail.contact_id}:notes",
                    label = t("thread.fieldNotes"),
                    initial = (contact?.notes ?: detail.contact.notes).orEmpty(),
                    maxLength = CONTACT_NOTES_MAX,
                    placeholder = t("thread.notesPlaceholder"),
                    singleLine = false,
                    save = { value -> controller.saveContactField("notes", value) },
                )
            }

            SheetSection(t("thread.sectionConsent")) {
                Text(
                    consentLine(
                        consentSource = contact?.consent_source
                            ?: detail.contact.consent_source,
                        consentAt = contact?.consent_at ?: detail.contact.consent_at,
                        consentAttestedBy = contact?.consent_attested_by,
                        memberName = ::memberName,
                    ),
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }

            // #301: where this customer came from. Above the tasks because it
            // is a question somebody ASKS in the first minute of a call, while
            // a task is what they write down afterwards — and because the ask
            // disappears the moment it is answered.
            SheetSection(t("thread.sectionLeadSource")) {
                LeadSourcePicker(controller = controller, detail = detail)
            }

            SheetSection(t("thread.sectionTasks")) {
                TasksChecklist(
                    state = controller.conversationTasks,
                    onToggle = { controller.toggleTaskDone(it) },
                    onOpenTask = onOpenTask,
                )
            }

            SheetSection(t("thread.sectionOtherConversations")) {
                OtherConversations(
                    state = controller.otherConversations,
                    onOpen = onOpenConversation,
                )
            }

            Spacer(Modifier.height(24.dp))
        }
    }
}

/**
 * #301 — "how did you hear about us?", as one tap.
 *
 * Hand-port of `apps/web/src/components/contact-panel/lead-source-picker.tsx`.
 *
 * #301's devil's-advocate section names the trap this is built around: asking
 * the tech to categorise every inbound is a tax on the person with the least
 * time, and if it is not one tap it will not happen — which produces a source
 * field empty 80% of the time and a MISLEADING report rather than no report.
 * So it is chips and not a menu.
 *
 * IT NEVER ASKS A QUESTION IT ALREADY KNOWS THE ANSWER TO. When the LINE
 * attributed the conversation there is nothing to ask, so it states the answer
 * and offers no prompt; asking anyway is how a crew learns to dismiss this
 * control, and the whole value of per-number attribution is that nobody has to
 * do anything.
 */
@OptIn(ExperimentalLayoutApi::class)
@Composable
private fun LeadSourcePicker(controller: ThreadController, detail: ConversationDetail) {
    val scope = rememberCoroutineScope()
    var sources by remember { mutableStateOf<List<LeadSource>>(emptyList()) }
    var pending by remember { mutableStateOf(false) }
    var current by remember(detail.id) { mutableStateOf(detail.lead_source_id) }
    var origin by remember(detail.id) { mutableStateOf(detail.lead_source_origin) }

    LaunchedEffect(detail.id) {
        // A list that will not load hides the picker rather than showing a
        // prompt with no answers on offer.
        sources = runCatching { controller.leadSources() }.getOrDefault(emptyList())
    }

    val options = sources.filter { it.archived_at == null }
    if (options.isEmpty()) return

    // An archived source still NAMES the thread it attributed — this
    // conversation genuinely came from the yard sign, even after it came down.
    val currentName = sources.firstOrNull { it.id == current }?.name

    fun choose(id: String?) {
        pending = true
        scope.launch {
            try {
                val next = controller.setLeadSource(detail.id, id)
                current = next.lead_source_id
                origin = next.lead_source_origin
            } finally {
                pending = false
            }
        }
    }

    Column(Modifier.fillMaxWidth()) {
        when {
            origin == "number" && currentName != null -> Text(
                t("thread.leadFromLine", "name" to currentName),
                style = MaterialTheme.typography.bodyMedium,
            )
            currentName != null -> Text(
                t("thread.leadSaidSo", "name" to currentName),
                style = MaterialTheme.typography.bodyMedium,
            )
            else -> Text(
                t("thread.leadAsk"),
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
        FlowRow(
            Modifier.fillMaxWidth().padding(top = 6.dp),
            horizontalArrangement = Arrangement.spacedBy(6.dp),
        ) {
            options.forEach { source ->
                val selected = source.id == current
                FilterChip(
                    selected = selected,
                    enabled = !pending,
                    // Tapping the chosen one again clears it: the fastest way
                    // back from a mistap is the control you just used.
                    onClick = { choose(if (selected) null else source.id) },
                    label = { Text(source.name) },
                )
            }
            if (current != null) {
                TextButton(enabled = !pending, onClick = { choose(null) }) {
                    Text(t("thread.dontKnow"))
                }
            }
        }
    }
}

@Composable
private fun SheetSection(title: String, content: @Composable () -> Unit) {
    Column(Modifier.fillMaxWidth()) {
        Text(
            title,
            style = MaterialTheme.typography.labelMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            modifier = Modifier.padding(bottom = 6.dp),
        )
        content()
    }
}

@Composable
private fun TasksChecklist(
    state: LoadState<List<Task>>?,
    onToggle: (Task) -> Unit,
    onOpenTask: ((taskId: String) -> Unit)?,
) {
    when (state) {
        null, is LoadState.Loading -> LoadingIndicator()
        is LoadState.Failed -> Text(
            t("thread.tasksLoadFailed"),
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )

        is LoadState.Ready -> {
            if (state.value.isEmpty()) {
                Text(
                    t("thread.noTasks"),
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
                return
            }
            Column {
                state.value.forEach { task ->
                    Row(
                        Modifier
                            .fillMaxWidth()
                            // #217: tapping the ROW opens the task detail; the
                            // Checkbox below stays a SEPARATE hit target so a
                            // done toggle never navigates.
                            .let { base ->
                                if (onOpenTask != null) base.clickable { onOpenTask(task.id) }
                                else base
                            },
                        verticalAlignment = Alignment.CenterVertically,
                    ) {
                        Checkbox(
                            checked = task.done,
                            onCheckedChange = { onToggle(task) },
                        )
                        Text(
                            task.title,
                            style = MaterialTheme.typography.bodyMedium,
                            textDecoration = if (task.done) TextDecoration.LineThrough else null,
                            color = if (task.done) {
                                MaterialTheme.colorScheme.onSurfaceVariant
                            } else {
                                MaterialTheme.colorScheme.onSurface
                            },
                            maxLines = 2,
                            overflow = TextOverflow.Ellipsis,
                            modifier = Modifier.weight(1f),
                        )
                    }
                }
            }
        }
    }
}

@Composable
private fun OtherConversations(
    state: LoadState<List<com.loonext.android.core.model.ConversationListItem>>?,
    onOpen: ((String) -> Unit)?,
) {
    when (state) {
        null, is LoadState.Loading -> LoadingIndicator()
        is LoadState.Failed -> Text(
            t("thread.priorLoadFailed"),
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )

        is LoadState.Ready -> {
            if (state.value.isEmpty()) {
                Text(
                    t("thread.noOtherConversations"),
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
                return
            }
            Column {
                state.value.forEach { row ->
                    Row(
                        Modifier
                            .fillMaxWidth()
                            .let { base ->
                                if (onOpen != null) base.clickable { onOpen(row.id) }
                                else base
                            }
                            .padding(vertical = 8.dp),
                        verticalAlignment = Alignment.CenterVertically,
                    ) {
                        Column(Modifier.weight(1f)) {
                            Text(
                                row.last_message?.body?.ifBlank { null }
                                    ?: t("thread.conversationFallback"),
                                style = MaterialTheme.typography.bodyMedium,
                                maxLines = 1,
                                overflow = TextOverflow.Ellipsis,
                            )
                            Text(
                                statusLabel(row.status),
                                style = MaterialTheme.typography.labelSmall,
                                color = MaterialTheme.colorScheme.onSurfaceVariant,
                            )
                        }
                        Spacer(Modifier.width(8.dp))
                        Text(
                            relativeTime(row.last_message_at),
                            style = MaterialTheme.typography.labelSmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                        )
                    }
                    HorizontalDivider(color = MaterialTheme.colorScheme.outlineVariant)
                }
            }
        }
    }
}
